[CHERI] Add a sealed type annotation.

resistor opened the pull request:
This allows us to disallow dereferences and pointer indexing on
sealed capabilities at compile time.

resistor committed:
[CHERI] Add a sealed type annotation.

This allows us to disallow dereferences and pointer indexing on
sealed capabilities at compile time.

davidchisnall commented:
Writing up discussion from elsewhere:

I think the builtins may need to preserve the type.  The biggest benefit of this is being able to do a `T*sealed` conversion to a possibly-null `T*`.  In C++, we can write a wrapper that is something like:

```c++
template<typename T>
T *cheri_unseal(T *__sealed sealed, void *sealingKey)
{
  return static_cast<T*>(__builtin_cheri_unseal(sealed, sealingKey));
}
```

In C, I'm not sure this is possible.  The following might work?

```
#define cheri_unseal(sealed, sealingKey) \
  ((__typeof__(*sealed)*) __builtin_cheri_unseal(sealed, sealingKey))
```

Though this would have to allow dereferencing of `sealed` in unevaluated contexts (we probably want that anyway, so that `sizeof(*sealed)` works). 

If this works then the builtins can work purely on `void*`.

resistor commented:
> Though this would have to allow dereferencing of sealed in unevaluated contexts (we probably want that anyway, so that sizeof(*sealed) works).

I *believe* this works, but would have to check.

resistor commented:
> > Though this would have to allow dereferencing of sealed in unevaluated contexts (we probably want that anyway, so that sizeof(*sealed) works).
> 
> I _believe_ this works, but would have to check.

Nope, it doesn't:

> foo.c:2:19: error: invalid operator applied to sealed type 'void * __sealed_capability'
>     2 |     return sizeof(*a);

To get that working, we'd need an implementation strategy more like `[[clang::noderef]]`. That annotation handles unevaluated contexts, but it operates strangely from our perspective: it annotated the pointee type rather than the pointer, and then warns if a value of pointee type is ever manipulated.

resistor committed:
[CHERIoT] Support most operations on __sealed_capabilities in unevaluated contexts.

This is needed in order to support things like sizeof and decltype.
Casts to an from sealed types are still disallowed.

resistor commented:
Supporting unevaluated contexts ended up being easier than I expected!

Now the cases that `[[clang::noderef]]`` allows and we don't are mostly limited to things like `&arr[0]` in evaluated contexts.

davidchisnall commented:
This appears to be missing versions of several of the builtins for sealed things.  For example, `__builtin_cheri_tag_get` now fails on sealed things.  Worse, `__builtin_cheri_type_get` also fails (and is mostly meaningless on unsealed things: unsealed things always have a type of 0).

davidchisnall commented:
And it looks as if `reinterpret_cast` can't cast away the sealed qualifier either.

davidchisnall commented:
Neither can C-style casts.  I'm currently working around this by casting via `uintptr_t`, which is very ugly.

We also have code that relies on being able to implicitly cast any pointer to `const void*` or `const volatile void*`.  I'm not sure what the right thing to do here is (@nwf, maybe you have opinions?).  On the one hand, it's nice that a `void*` doesn't let you accidentally lose sealed, but on the other hand we often use `void*` as an 'any pointer' thing, and it's annoying to have to say 'any pointer except a sealed one' and have a special case for sealed.  I think I lean towards thinking you should be able to implicitly cast `T *__sealed_capability` to `void*`, since casting *from* `void*` (at least in C++, not in C) requires an explicit cast, and anything casting from `void*` should know what it's doing.

davidchisnall commented:
Superseded by #109
